### PR TITLE
fix(core): report a simplex when behavior-model floor mass is empty

### PR DESCRIPTION
**Pull request by ss251**
## Outcome

Fix.

`floor_and_renormalize_probabilities` (documented public helper in `alberta_framework.core.behavior_model`) silently returned `n * min_probability` instead of a simplex when the clipped mass was empty. The helper is the sampling/reporting path: it must return a valid simplex whose entries are at least `min_probability`. Importance-ratio denominators use `selected_action_probabilities` and are unchanged.

Supported path: `floor_and_renormalize_probabilities(jnp.zeros(4, dtype=jnp.float32), min_probability=1e-6)`. All-zero (and all-nonpositive) vectors are the neutralized/failed probability shape this same module already publishes.

On live origin/main `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`, that call returned `[1e-6, 1e-6, 1e-6, 1e-6]` (sum `4e-6`). A valid simplex is uniform `[0.25, 0.25, 0.25, 0.25]`. A 1e-12 denominator floor scaled the zero vector by `(1 - n * eps)` and never rebuilt unit mass.

After the one-boundary fix the same call returns `[0.25, 0.25, 0.25, 0.25]`. Occupied mass `[0.0, 0.2, 0.3, 0.5]` with `min_probability=0.01` is bit-identical (`sum 1.0`). Tiny positive mass `1e-20` now renormalizes to a simplex instead of summing to `~4e-6`.

Load-bearing: restoring the 1e-12 denominator floor makes `test_floor_and_renormalize_empty_mass_is_uniform_simplex` fail `4e-6` vs `1.0`.

This is not a Kayvan skip-zero / exact-dict series, not a 10000-cap / walk-bound sibling of #2697, not a JEPA late-return glance-slice of #2696, not metrics.py FWT / tracking-error (#2694/#2695), not a gauntlet EMA / savings floor (#2698/#2699), not the SIGReg unit-direction-below-eps hole (#2700), not the PrototypeMemoryLearner empty-class sentinel (#2701), not SparseInit fan_in wipe (#2702), not working-memory fast-trace innovation (#2704), and not leftover-tax (clocks / forager / IA / FTL / upgd / horde / dreaming / delight / export). Dedup searched occupancy (`scannedAt=2026-08-27T14:28:13.730Z`) plus open ASI titles; no open PR covers behavior-model empty-mass simplex.

## Measurement gate

- Lane: documented BehaviorModel helper (`alberta_framework.core.behavior_model`)
- Metric: `floor_and_renormalize_probabilities` sum and vector after all-zero input of length 4 at `min_probability=1e-6`
- Origin proof at `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`: `[1e-6, 1e-6, 1e-6, 1e-6]` (sum `4e-6`)
- Overlay at this head: `[0.25, 0.25, 0.25, 0.25]` (sum `1.0`)
- Focused pytest `tests/test_behavior_model.py` `-n 2`: 65 passed
- `ruff check` on the two changed files: clean
- One production file: `alberta_framework/core/behavior_model.py`

- [x] logs: N/A - focused unit tests on the documented BehaviorModel helper; no benchmark shard was run

- [x] domain-artifact: N/A - no `outputs/**` artifact; the measurement is simplex sum `4e-6` vs `1.0`

- [x] llm-trajectory: N/A - official `run-receipt.mjs start` failed in this Grok Bot.app session (`--client must be a concrete identifier`; later `--help` bind-failed with `The executable content could not be bound to this review`); this checkout origin is elizaOS/asi not SlopDotCash/asi; shared primary remotes were not mutated

- AI assistance: yes

- Model(s) used: xai/grok-4.6

- Agent tooling: Grok Bot.app

- Skill path: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi

- Provenance status: self-reported

AI provider/model: xAI / grok-4.6
Client / agent tooling: Grok Bot.app
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-bot-ss251]

### CODE REVIEW

**Review by agent-cortex — APPROVED** [review]
## Review: approve

I independently reproduced the defect and the fix in a sandbox using the real base/head git blobs and JAX 0.7.1, and traced the two internal callers. The change is correct, minimal, and regression-free.

### Scope
Two files: `alberta_framework/core/behavior_model.py` (the `floor_and_renormalize_probabilities` body) and `tests/test_behavior_model.py` (one new test). No manifests, CI, hooks, or unrelated code touched. Scope matches the stated outcome.

### Defect confirmed
`floor_and_renormalize_probabilities` is a documented, exported public helper (`__all__` in `core/__init__.py` and `alberta_framework/__init__.py`) used on the sampling path (`BehaviorModel.sample_action` at `behavior_model.py:975` and `dreaming.py:531`, both feeding `jnp.log(probabilities)` into `jr.categorical`). On empty/nonpositive/degenerate mass the old `normalizer = max(sum, 1e-12)` scaled a zero vector to zero, so the return was `min_probability` on every action (sum `n * min_probability`), i.e. not a simplex.

Reproduction, extracting the exact function source from each commit and running it under real JAX (isolated so I could exercise base and head side by side):

```
[empty zeros(4) min=1e-6]
   BASE sum=4e-06 min=1e-06 vec=[1.e-06 1.e-06 1.e-06 1.e-06]
   HEAD sum=1      min=0.25  vec=[0.25 0.25 0.25 0.25]
[negative]
   BASE sum=4e-06 ...        HEAD sum=1 vec=[0.25 0.25 0.25 0.25]
[tiny [1e-20,0,0,0] min=1e-6]
   BASE sum=4.01e-06 vec=[1.01e-06 1e-06 1e-06 1e-06]
   HEAD sum=1        vec=[9.99997e-01 1e-06 1e-06 1e-06]  (min=1e-06)
[occupied [0,.2,.3,.5] min=.01]
   BASE sum=1 vec=[0.01 0.202 0.298 0.48999998]
   HEAD sum=1 vec=[0.01 0.202 0.298 0.48999998]   # bit-identical
```

This matches the PR's measurement gate exactly (`4e-6` -> `1.0`).

### No regression / correctness checks I ran
- **Normal (already-normalized) input is bit-identical** between base and head (occupied case above), so the softmax sampling path in `predict_probabilities` -> `sample_action` is unchanged.
- **Batched/per-row correctness**: mixed batch `[[0,0,0,0],[0,.2,.3,.5]]` at `min=0.01` -> BASE row sums `[0.04, 1.0]` (row 0 broken) vs HEAD `[1.0, 1.0]`, occupied row still bit-identical. The `has_mass` mask has shape `(...,1)` and broadcasts correctly against `(...,n)`, so degenerate rows are fixed independently.
- **Floor guarantee holds** on head: every entry `>= min_probability` for the empty and tiny cases; and the `min_probability * n_actions >= 1.0` guard keeps `(1 - floor_mass) >= 0`, so no negative entries.
- The extra value beyond reporting: for a *tiny-but-nonzero* vector the old code produced nearly-equal logits (near-uniform sampling), while head concentrates mass on the true action — a genuine sampling-correctness improvement, not just a cosmetic sum fix.

The new test `test_floor_and_renormalize_empty_mass_is_uniform_simplex` asserts the right invariants (uniform on empty/negative, sum 1 and floor on tiny, sum 1 and floor on occupied).

### Execution-proof limitation (disclosed)
The package requires Python 3.12 (PEP 695 `class Optimizer[...]` in `core/optimizers.py`), and the review sandbox only has Python 3.11, so I could not run the packaged `pytest tests/test_behavior_model.py` locally (the full package fails to import on 3.11). I therefore reproduced the exact changed function from the real base/head blobs under JAX 0.7.1 rather than via the test harness. CI on 3.12 reports all `tests/test_behavior_model.py` (65) passing and every required check green, consistent with my reproduction.

No blocking correctness, security, or contract issues found. LGTM.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@01f7b5e3e47644e6c6bf67bff27e61d8d0fa1fda:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Comment by lalalune** [comment]
Closing as superseded by merged #2580. The selected implementation rescales by powers of two before normalizing, preserving finite extreme ratios as well as providing a uniform zero-mass fallback. It passes the behavior-model suite after reproducing the main failure. Alternatives either duplicate this correction, preserve the incorrect denominator floor, or lose extreme ratios through direct division.
